lib/internal/modules/run_main.js

```diff
@@ -70,7 +72,19 @@ function shouldUseESMLoader(mainPath) {
   const { readPackageScope } = require('internal/modules/package_json_reader');
   const pkg = readPackageScope(mainPath);
   // No need to guard `pkg` as it can only be an object or `false`.
-  return pkg.data?.type === 'module' || getOptionValue('--experimental-default-type') === 'module';
+  switch (pkg.data?.type) {
+    case 'module':
+      return true;
+    case 'commonjs':
+      return false;
+    default: { // No package.json or no `type` field.
+      if (getOptionValue('--experimental-detect-module')) {
+        const content = fs.readFileSync(mainPath, 'utf8');
+        return containsModuleSyntax(content, mainPath);
```

Comment:
There is an optimization opportunity in here. No need to call readFileSync. We can call it directly from `containsModuleSyntax` function to improve the performance. If we are not going to address it in this pull request, I recommend adding a TODO in here.

Reply:
[`5b6bdb4` (#50096)](https://github.com/nodejs/node/pull/50096/commits/5b6bdb41c8fbceca30c181cca9a9e95e675d942e)